Microsoft.Toolkit.Uwp.UI.Controls/WrapPanel/WrapPanel.cs

```diff
@@ -0,0 +1,157 @@
+        protected override Size MeasureOverride(Size availableSize)
+        {
+            _vSize = new List<double> { 0.0 };
+            var totalMeasure = new UvMeasure();
+            var parentMeasure = new UvMeasure(Orientation, availableSize.Width, availableSize.Height);
+            var lineMeasure = new UvMeasure();
+            foreach (var child in Children)
+            {
+                child.Measure(availableSize);
+
+                var currentMeasure = new UvMeasure(Orientation, child.DesiredSize.Width, child.DesiredSize.Height);
+
+                if (parentMeasure.U > currentMeasure.U + lineMeasure.U)
+                {
+                    lineMeasure.U += currentMeasure.U;
+                    lineMeasure.V = Math.Max(lineMeasure.V, currentMeasure.V);
+                }
+                else
+                {
+                    // new line should be added
+                    // to get the max U to provide it correctly to ui width ex: ---| or -----|
+                    totalMeasure.U = Math.Max(lineMeasure.U, totalMeasure.U);
+                    totalMeasure.V += lineMeasure.V;
+                    _vSize.Add(lineMeasure.V);
+
+                    // if the next new row still can handle more controls
+                    if (parentMeasure.U > currentMeasure.U)
+                    {
+                        // set lineMeasure initial values to the currentMeasure to be calculated later on the new loop
+                        lineMeasure = currentMeasure;
+                    }
+
+                    // the control will take one row alone
+                    else
+                    {
+                        // validate the new control measures
+                        totalMeasure.U = Math.Max(currentMeasure.U, totalMeasure.U);
+                        totalMeasure.V += currentMeasure.V;
+                        _vSize.Add(currentMeasure.V);
+
+                        // add new empty line
+                        lineMeasure = new UvMeasure();
+                    }
+                }
+            }
+
+            // update value with the last line
+            // if the the last loop is(parentMeasure.U > currentMeasure.U + lineMeasure.U) the total isn't calculated then calculate it
+            // if the last loop is (parentMeasure.U > currentMeasure.U) the currentMeasure isn't added to the total so add it here
+            // for the last condition it is zeros so adding it will make no difference
+            // this way is faster than an if condition in every loop for checking the last item
+            totalMeasure.U = Math.Max(lineMeasure.U, totalMeasure.U);
+            totalMeasure.V += lineMeasure.V;
+            _vSize.Add(lineMeasure.V);
+
+            return Orientation == Orientation.Horizontal ? new Size(totalMeasure.U, totalMeasure.V) : new Size(totalMeasure.V, totalMeasure.U);
+        }
+
+        /// <inheritdoc />
+        protected override Size ArrangeOverride(Size finalSize)
```

Comment:
I think this can be simplified by removing the _vSize variable:
```
protected override Size ArrangeOverride(Size finalSize)
{
    var parentMeasure = new UvMeasure(Orientation, finalSize.Width, finalSize.Height);
    var position = new UvMeasure();

    double currentV = 0;
    foreach (var child in Children)
    {
        var desiredMeasure = new UvMeasure(Orientation, child.DesiredSize.Width, child.DesiredSize.Height);
        if ((desiredMeasure.U + position.U) > parentMeasure.U)
        {
            // next row!
            position.U = 0;
            position.V += currentV;
            currentV = 0;
        }

        // Place the item
        if (Orientation == Orientation.Horizontal)
        {
            child.Arrange(new Rect(position.U, position.V, child.DesiredSize.Width, child.DesiredSize.Height));
        }
        else
        {
            child.Arrange(new Rect(position.V, position.U, child.DesiredSize.Width, child.DesiredSize.Height));
        }

        // adjust the location for the next items
        position.U += desiredMeasure.U;
        currentV = Math.Max(desiredMeasure.V, currentV);
    }
    return finalSize;
}
```